### PR TITLE
Add support for HttpChecksum trait - support flexible checksums.

### PR DESCRIPTION
**Pull request by alextwoods**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/version-3/CONTRIBUTING.md

Thank you for your contribution!

### CODE REVIEW

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+      # @api private
+      class OptionHandler < Seahorse::Client::Handler
+        def call(context)
+          context[:http_checksum] ||= {}
+
+          # validate request configuration
+          if (request_input = ChecksumAlgorithm.request_algorithm_selection(context))
+            unless CLIENT_ALGORITHMS.include? request_input
+              raise ArgumentError, "#{request_input} is not a supported checksum algorithm." # TODO: check for CRC32C and add specific crt help
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb]: Still necessary to do?

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+      # @api private
+      class ChecksumHandler < Seahorse::Client::Handler
+
+        def call(context)
+          if should_calculate_request_checksum?(context)
+            request_algorithm_input = ChecksumAlgorithm.request_algorithm_selection(context)
+            context[:checksum_algorithms] = request_algorithm_input
+
+            request_checksum_property =           {
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb]: Weird spacing

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+          http_response.on_data do |chunk|
+            checksum_context[:digest].update(chunk) if checksum_context[:digest]
+          end
+
+          http_response.on_success do
+            if checksum_context[:digest] &&
+              (computed = checksum_context[:digest].base64digest)
+
+              if computed != checksum_context[:expected]
+                raise Aws::Errors::ChecksumError, "Checksum validation failed on #{checksum_context[:header_name]} computed: #{computed}, expected: #{checksum_context[:expected]}"
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb]: Wrap

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+          @io.rewind
+        end
+
+        def read(length, buf)
+          # account for possible leftover bytes at the end, if we have trailer bytes, send them
+          if @trailer_io
+            return @trailer_io.read(length, buf)
+          end
+
+            chunk = @io.read(length)
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb]: Bad indenting

**Review by mullermp — APPROVED** [review]
Looks good. Are all SEP cases covered? Have we covered all abstractions (resource models?)

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb (hunk 1 above)
Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb]: Yeah - updated to add CRT instructions.

**Pull request merged**